### PR TITLE
HIVE-26437: Dump unpartitioned tables in parallel

### PR DESCRIPTION
**Pull request by atsaonerk**
HIVE-26437: Dump unpartitioned tables in parallel.

Currently partitions of table is dump in parallel manner. But if table is not partitioned, it is dumped serially.
This change introduces parallelism at table level as well. A single thread pool which is currently being used
for partition level is also used for table level. The table level dump task is added to same thread pool.
The degree of parallelism depends upon config parameter REPL_PARTITIONS_DUMP_PARALLELISM whose defaul value is 100.
The new ExportService is introduced with this change which would be responsible for exporting table and partitions during repl dump. The ExportService is initialized and configured with thread pools by HiveServer2 service.
A new Hiveconfig variable ie "REPL_TABLE_DUMP_PARALLELISM is introduced to define the number of threads which would be created in thread pool. 
The ExportService which is created as singleton instance would be used by ReplDumpTask.

### CODE REVIEW

**Comment by cmunkey** [comment]
Coding standard is 2 space indents.

The unit test mainly tests the DumpExporter since the runnable does nothing.
Maybe there needs to be an integration test that makes sure that running exports in parallel doesn't
have issues like more than one thread accessing some resource that we don't know about.

For your existing test, maybe the multithreaded one can do a runnable like this:

- counter that is a semaphore set to the number of currency that you want
- each runnable, 
     counter.tryAcquire() - should not block (return false)
     sleep
     counter.release()

**Review by cmunkey — COMMENTED** [review]
Need to test the test too.

Add a test like this:

public void testDumpExporterWithParallelismFail() throws Exception {
   when(conf.getIntVar(HiveConf.ConfVars.REPL_EXPORT_DUMP_PARALLELISM)).thenReturn(10);
    sem = new Semaphore(totalTask-1);
    
    
This test should fail intentionally.
So you will need to catch the exception and PASS if it fails.
I think there is an annotation for that like

@Test(expected=NullPointerException.class)

replace with your exception

**Review by cmunkey — COMMENTED** [review]
Needs an integration test to verify that these operations can indeed run in parallel
w/o concurrency issues accessing resources.

**Comment by atsaonerk** [comment]
> Need to test the test too.
> 
> Add a test like this:
> 
> public void testDumpExporterWithParallelismFail() throws Exception { when(conf.getIntVar(HiveConf.ConfVars.REPL_EXPORT_DUMP_PARALLELISM)).thenReturn(10); sem = new Semaphore(totalTask-1);
> 
> This test should fail intentionally. So you will need to catch the exception and PASS if it fails. I think there is an annotation for that like
> 
> @test(expected=NullPointerException.class)
> 
> replace with your exception

What is the failure expected?
We have for e.g 7 semaphore limits and executing 8 runnable task.
The 8th one wont get semaphore and return without incrementing task number.
The ASSURE_TRUE(7,8) will fail. Is this the one you are saying?

**Review by cmunkey — COMMENTED** [review]
> @Test(expected = HiveException.class)
>  public void testDumpExporterWithParallelismFail() throws Exception {

Shouldn't the error be AssertionError?
Is what's failing really what you expected?

I think this is what's wrong:
> Assert.assertTrue(sem.tryAcquire());

Should be different for positive vs negative test.

Positive tests should be true,
Negative tests should be false.

Are all your other tests failing too? they may, you are catching and not rethrowing there.

This should be what's failing. assertTrue() throw

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosExternalTables.java [hunk 1]
```diff
@@ -1937,4 +1937,84 @@ public void testSingleCopyTasksConfiguration(boolean runCopyTasksOnTarget)
...
+  public void testTableExportDumpParallelism(boolean enableParallelism) throws Throwable {
+    List<String> tableList = new ArrayList<String>();
+    List<String> withClause = ReplicationTestUtils.includeExternalTableClause(true);
+    if (enableParallelism) {
+      withClause.add("'" + HiveConf.ConfVars.REPL_EXPORT_DUMP_PARALLELISM.varname + "'='100'");
+    }
+
+    primary.run("use " + primaryDbName);
+
+    // create 10 partitioned tables.
```
Comment (cmunkey) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosExternalTables.java:1960]: Can you add managed tables as well.

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosExternalTables.java [hunk 2]
```diff
@@ -1937,4 +1937,84 @@ public void testSingleCopyTasksConfiguration(boolean runCopyTasksOnTarget)
...
+  @Test
+  public void testTableAndPartitionExportDumpWithNoParallelism() throws Throwable {
+    testTableExportDumpParallelism(false);
+  }
+
+  public void testTableExportDumpParallelism(boolean enableParallelism) throws Throwable {
+    List<String> tableList = new ArrayList<String>();
+    List<String> withClause = ReplicationTestUtils.includeExternalTableClause(true);
+    if (enableParallelism) {
+      withClause.add("'" + HiveConf.ConfVars.REPL_EXPORT_DUMP_PARALLELISM.varname + "'='100'");
```
Comment (cmunkey) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosExternalTables.java:1955]: Should also test where there are less threads than tables and partitions.
Especially the cases where the number of threads is less than the number of tables and number of partitions.
For instance 10 threads, 20 tables, 4th and 5th tables have > 10 partitions.

File: ql/src/test/org/apache/hadoop/hive/ql/parse/repl/dump/TestDumpExporter.java [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+  HiveConf conf;
+
+  private static int taskNumber = 0;
+
+  private final int totalTask = 10;
+
+  private Semaphore sem;
+
+  public void runParallelTask() {
+    Assert.assertTrue(sem.tryAcquire());
```
Comment (cmunkey) [inline ql/src/test/org/apache/hadoop/hive/ql/parse/repl/dump/TestDumpExporter.java:53]: This should fail with AssertionException in the case where you have less semaphores than threads.

File: ql/src/test/org/apache/hadoop/hive/ql/parse/repl/dump/TestDumpExporter.java [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+      throw new RuntimeException(e);
+    }
+  }
+
+  public void runSequentialTask() {
+    ++taskNumber;
+    LOG.debug("Current task number is: {} and thread is: {} ", taskNumber, Thread.currentThread().getName());
+  }
+
+  protected void finalize()
```
Comment (cmunkey) [inline ql/src/test/org/apache/hadoop/hive/ql/parse/repl/dump/TestDumpExporter.java:68]: You can make this some \@After or \@AfterClass.

File: ql/src/test/org/apache/hadoop/hive/ql/parse/repl/dump/TestDumpExporter.java [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+    DumpExporter dumpExporter = new DumpExporter(conf);
+    try {
+      for (int i = 0; i < totalTask; i++) {
+        dumpExporter.submit(r);
+      }
+      final long actualNumTasksExecuted = dumpExporter.getTotalTaskEverExecuted();
+      Assert.assertEquals(totalTask, actualNumTasksExecuted);
+      dumpExporter.finish();
+    } catch (Exception e) {
+      LOG.info("DumpExporter task failed with exception {} ", e.getMessage());
```
Comment (cmunkey) [inline ql/src/test/org/apache/hadoop/hive/ql/parse/repl/dump/TestDumpExporter.java:91]: Your test will still pass if there is an exception. Why are you catching here?

File: ql/src/test/org/apache/hadoop/hive/ql/parse/repl/dump/TestDumpExporter.java [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+    };
+    DumpExporter dumpExporter = new DumpExporter(conf);
+    try {
+      for (int i = 0; i < totalTask; i++) {
+        dumpExporter.submit(r);
+      }
+      final long actualNumTasksExecuted = dumpExporter.getTotalTaskEverExecuted();
+      Assert.assertEquals(0, actualNumTasksExecuted);
+      dumpExporter.finish();
+    } catch (Exception e) {
```
Comment (cmunkey) [inline ql/src/test/org/apache/hadoop/hive/ql/parse/repl/dump/TestDumpExporter.java:111]: Test will pass if there is an exception, why are you catching here.

File: ql/src/test/org/apache/hadoop/hive/ql/parse/repl/dump/TestDumpExporter.java [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+      LOG.info("DumpExporter task failed with exception {} ", e.getMessage());
+    }
+  }
+
+  @Test(expected = HiveException.class)
```
Comment (cmunkey) [inline ql/src/test/org/apache/hadoop/hive/ql/parse/repl/dump/TestDumpExporter.java:116]: I would expect AssertionException. How are you getting HiveException?

**Comment by cmunkey** [comment]
Ignore these comments, I added them in line instead.

> > @test(expected = HiveException.class)
> > public void testDumpExporterWithParallelismFail() throws Exception {
> 
> Shouldn't the error be AssertionError? Is what's failing really what you expected?
> 
> I think this is what's wrong:
> 
> > Assert.assertTrue(sem.tryAcquire());
> 
> Should be different for positive vs negative test.
> 
> Positive tests should be true, Negative tests should be false.
> 
> Are all your other tests failing too? they may, you are catching and not rethrowing there.
> 
> This should be what's failing. assertTrue() throw

File: ql/src/java/org/apache/hadoop/hive/ql/parse/repl/dump/DumpExporter.java [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+          LOG.error("DumpExporter task failed with reason: {} ", e.getCause());
+          throw new HiveException(e.getMessage(), e);
+        }
+      }
+    }
+  }
+
+  public long getTotalTaskEverExecuted() {
+    if (!isSingleThreadedDumpExporter()) {
+      ThreadPoolExecutor executor = (ThreadPoolExecutor) execService;
```
Comment (pudidic) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/repl/dump/DumpExporter.java:88]: How about cast in advance? Such as; private ThreadPoolExecutor executor;

File: ql/src/java/org/apache/hadoop/hive/ql/parse/repl/dump/DumpExporter.java [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+import org.apache.hadoop.hive.metastore.ThreadPool;
+import org.apache.hadoop.hive.ql.metadata.HiveException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.LinkedList;
+import java.util.List;
+import java.util.concurrent.*;
+
+public class DumpExporter {
```
Comment (pudidic) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/repl/dump/DumpExporter.java:32]: The role of DumpExporter isn't to export, which its name implies. It's a wrapper of an ExecutorService, which accepts general Runnable objects with the configured number of threads with a logger. Exposing DumpExporter.submit(Runnable) can lead to unexpected results, especially eating up the thread pool with unexpected tasks and delaying the expected ones.

I suggest two design change options;
1. Call private ExecutorService.submit() in a service only. See ScheduledQueryExecutionService and ScheduledQueryExecutor for its pattern. Rename DumpExporter as ExportService and move TableExport and PartitionExport inside ExportService. Hide the object creation from outside and make method call instead.
2. Publish submit(T) and expose interface T for future extension. See JobRequestExecutor and EventualCleanupService for its pattern. Rename DumpExporter as ExportExecutor. Introduce a new interface Export and limit submit(Export) method parameter type. Let TableExport and PartitionExport to implement Export interface. If you can find common methods, you can put them in the Export interface.

I prefer the first design as I see there's a small room for future extension. Also please add class description.

File: ql/src/java/org/apache/hadoop/hive/ql/parse/repl/dump/DumpExporter.java [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.apache.hadoop.hive.metastore.ThreadPool;
+import org.apache.hadoop.hive.ql.metadata.HiveException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.LinkedList;
+import java.util.List;
+import java.util.concurrent.*;
```
Comment (pudidic) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/repl/dump/DumpExporter.java:30]: Please import each classes explicitly.

**Comment by pudidic** [comment]
For performance benchmark, please see https://github.com/apache/hive/blob/master/itests/hive-jmh/src/main/java/org/apache/hive/benchmark/vectorization/VectorizedArithmeticBench.java . It's a JMH benchmark. It's simple and has good explanation on how to use it.

File: ql/src/java/org/apache/hadoop/hive/ql/parse/repl/dump/DumpExporter.java [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+
+public class DumpExporter {
+  private ExecutorService execService;
```
Comment (subhasisgorai) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/repl/dump/DumpExporter.java:33]: Please mark the attributes as final if we don't expect them to be changed over time.

File: ql/src/java/org/apache/hadoop/hive/ql/parse/repl/dump/PartitionExport.java [hunk 12]
```diff
@@ -75,64 +77,62 @@ class PartitionExport {
     this.callersSession = SessionState.get();
   }
 
-  List<DataCopyPath> write(final ReplicationSpec forReplicationSpec, boolean isExportTask,
+  public void exportPartitions(final ReplicationSpec forReplicationSpec, boolean isExportTask,
+                            FileList fileList, boolean dataCopyAtLoad) throws RuntimeException {
```
Comment (pudidic) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/repl/dump/PartitionExport.java:81]: Please introduce an additional prefix and keep the same base name. Both of backing and wrapping methods share a same logical behavior, but in a different thread model. For example, parallelXxx(...) and xxx(...) can help readers to what they represent. Additionally, it's better to have a Javadoc document to explicitly say that parallelXxx is a parallelized version of xxx. You can use @ see annotation for its reference.

exportPartitions and write may lead readers to think that they do different things, and exportPartitions doesn't give a good clue about thread models.

A same advice applies for exportTable method, too.

File: ql/src/test/org/apache/hadoop/hive/ql/parse/repl/dump/TestExportService.java [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+    sem = new Semaphore(totalTask);
+    ExportService exportService = ExportService.getInstance();
+    exportService.configure(conf, true);
+    for (int i = 0; i < totalTask; i++) {
+      exportService.submit(runParallelTask());
+    }
+    final long actualNumTasksExecuted = exportService.getTotalTaskEverExecuted();
+    Assert.assertEquals(totalTask, actualNumTasksExecuted);
+    exportService.shutdown();
+    exportService.await(Long.MAX_VALUE, TimeUnit.SECONDS);
```
Comment (pudidic) [inline ql/src/test/org/apache/hadoop/hive/ql/parse/repl/dump/TestExportService.java:92]: Please consider a test failure case with Long.MAX_VALUE waiting time. Can few minutes or shorter catch them?

File: ql/src/java/org/apache/hadoop/hive/ql/parse/repl/dump/PartitionExport.java [hunk 14]
```diff
@@ -75,64 +77,62 @@ class PartitionExport {
...
-      because producer is not terminated but it wont produce anything so queue will be empty and then we
-      should only wait for a specific time before continuing, as the next loop cycle will fail.
-       */
+    /*
+    This is removed using a poll because there can be a case where there partitions iterator is empty
+    but because both the producer and consumer are started simultaneously the while loop will execute
+    because producer is not terminated but it wont produce anything so queue will be empty and then we
+    should only wait for a specific time before continuing, as the next loop cycle will fail.
+     */
       Partition partition = queue.poll(1, TimeUnit.SECONDS);
```
Comment (pudidic) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/repl/dump/PartitionExport.java:127]: Please respect the original indentation. Sometimes your indentation is better. But other committers may have a different opinion. We don't want to make repetitive changes to grow. Also it's much cleaner on patch files, too.

**Review by pudidic — CHANGES_REQUESTED** [review]
It seems like this change makes some test failures. Please solve them.

**Comment by pudidic** [comment]
Also please update the link with a HIVE issue, not the cloudera link.

**Comment by atsaonerk** [comment]
Results of jmh performance benchmark test indicates an improvement in TableExport operation. If table is dumped in parallel manner instead of serial , the operation completes much faster giving. Below is the result seen when 500 tablexport operations are done both in serial and parallel manner.

Result "org.apache.hive.benchmark.ql.exec.TableAndPartitionExportBench.BaseBench.parallel":
  N = 5
  mean =    640.862 ?(99.9%) 113.354 ms/op

Result "org.apache.hive.benchmark.ql.exec.TableAndPartitionExportBench.BaseBench.serial":
  N = 5
  mean =  51697.322 ?(99.9%) 322.747 ms/op

Benchmark                                                                     Mode  Cnt      Score             Error   Units
TableAndPartitionExportBench.BaseBench.parallel    ss         5         640.862        ?         113.354  ms/op
TableAndPartitionExportBench.BaseBench.serial       ss          5        51697.322     ?          322.747  ms/op

**Comment by pudidic** [comment]
The test failed. Please check [TestReplicationScenarios](http://ci.hive.apache.org/blue/organizations/jenkins/hive-precommit/detail/PR-3444/19/tests).

**Comment by atsaonerk** [comment]
The latest change is corrupted. Created a new PR.

**Pull request closed without merging**